sdk/python/lib/pulumi/runtime/rpc.py

```diff
@@ -362,23 +373,64 @@ def do_resolve(value_fut: 'asyncio.Future',
     return resolvers
 
 
-def translate_output_properties(res: 'Resource', output: Any) -> Any:
+def translate_output_properties(res: 'Resource', output: Any, typ: Optional[type] = None) -> Any:
     """
     Recursively rewrite keys of objects returned by the engine to conform with a naming
     convention specified by the resource's implementation of `translate_output_property`.
 
+    Additionally, if output is a `dict` and `typ` is an output type, instantiate the output type,
+    passing the dict as an argument to the output type's __init__() method.
+
     If output is a `dict`, every key is translated using `translate_output_property` while every value is transformed
     by recursing.
 
     If output is a `list`, every value is recursively transformed.
 
     If output is a primitive (i.e. not a dict or list), the value is returned without modification.
     """
+
+    # Unwrap optional types.
+    typ = _types.unwrap_optional_type(typ) if typ else typ
+
     if isinstance(output, dict):
-        return {res.translate_output_property(k): translate_output_properties(res, v) for k, v in output.items()}
+        # Function called to lookup a type for a given key.
+        # The default always returns None.
+        get_type: Callable[[str], Optional[type]] = lambda k: None
+
+        if typ and _types.is_output_type(typ):
+            # If typ is an output type, get its types, so we can pass
+            # the type along for each property.
+            types = _types.output_type_types(typ)
+            get_type = lambda k: types.get(k) # pylint: disable=unnecessary-lambda
+        elif typ:
+            # If typ is a dict, get the type for its values, to pass
+            # along for each key.
+            origin = _types.get_origin(typ)
+            if origin is dict or Dict:
```

Comment:
Should we report an error if the origin is not a dict?

Reply:
In general, I wonder if we should error on a type mismatch.

Reply:
Yes, I was wondering about that too. I think we should. I'll address.